flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -367,8 +377,9 @@ public void flush() {
 			}
 			// if there is more then 1 buffer, we already notified the reader
 			// (at the latest when adding the second buffer)
-			notifyDataAvailable = !flushRequested && buffers.size() == 1 && buffers.peek().isDataAvailable();
+			notifyDataAvailable = mayNotifyDataAvailable() && buffers.peekFirst().isDataAvailable();
 			flushRequested = flushRequested || buffers.size() > 1 || notifyDataAvailable;
+			dataAvailableNotified = dataAvailableNotified || notifyDataAvailable;
```

Comment:
Refer to https://github.com/apache/flink/pull/11786/files#r410367542, we can remove `dataAvailableNotified = dataAvailableNotified || notifyDataAvailable;`